substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,770 @@
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The account cannot perform this operation.
+		NotAllowed,
+		/// An existing staker cannot perform this action.
+		AlreadyStaking,
+		/// Reward Destination cannot be same as `Agent` account.
+		InvalidRewardDestination,
+		/// Delegation conditions are not met.
+		///
+		/// Possible issues are
+		/// 1) Cannot delegate to self,
+		/// 2) Cannot delegate to multiple delegates,
+		InvalidDelegation,
+		/// The account does not have enough funds to perform the operation.
+		NotEnoughFunds,
+		/// Not an existing `Agent` account.
+		NotAgent,
+		/// Not a Delegator account.
+		NotDelegator,
+		/// Some corruption in internal state.
+		BadState,
+		/// Unapplied pending slash restricts operation on `Agent`.
+		UnappliedSlash,
+		/// `Agent` has no pending slash to be applied.
+		NothingToSlash,
+		/// Failed to withdraw amount from Core Staking.
+		WithdrawFailed,
+		/// Operation not supported by this pallet.
+		NotSupported,
+	}
+
+	/// A reason for placing a hold on funds.
+	#[pallet::composite_enum]
+	pub enum HoldReason {
+		/// Funds held for stake delegation to another account.
+		#[codec(index = 0)]
+		Delegating,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Funds delegated by a delegator.
+		Delegated { agent: T::AccountId, delegator: T::AccountId, amount: BalanceOf<T> },
+		/// Funds released to a delegator.
+		Released { agent: T::AccountId, delegator: T::AccountId, amount: BalanceOf<T> },
+		/// Funds slashed from a delegator.
+		Slashed { agent: T::AccountId, delegator: T::AccountId, amount: BalanceOf<T> },
+	}
+
+	/// Map of Delegators to their `Delegation`.
+	///
+	/// Implementation note: We are not using a double map with `delegator` and `agent` account
+	/// as keys since we want to restrict delegators to delegate only to one account at a time.
+	#[pallet::storage]
+	pub(crate) type Delegators<T: Config> =
+		CountedStorageMap<_, Twox64Concat, T::AccountId, Delegation<T>, OptionQuery>;
+
+	/// Map of `Agent` to their `Ledger`.
+	#[pallet::storage]
+	pub(crate) type Agents<T: Config> =
+		CountedStorageMap<_, Twox64Concat, T::AccountId, AgentLedger<T>, OptionQuery>;
+
+	/// This pallet is not currently written with the intention of exposing any calls. But the
+	/// functions defined in the following impl block should act as a good reference for how the
+	/// exposed calls would look like when exposed.
+	impl<T: Config> Pallet<T> {
+		/// Register an account to become a stake `Agent`. Sometimes also called a `Delegatee`.
+		///
+		/// Delegators can authorize `Agent`s to stake on their behalf by delegating their funds to
+		/// them. The `Agent` can then use the delegated funds to stake to [`Config::CoreStaking`].
+		///
+		/// Implementation note: This function allows any account to become an agent. It is
+		/// important though that accounts that call [`StakingUnsafe::virtual_bond`] are keyless
```

Comment:
> It is important though that accounts that call [`StakingUnsafe::virtual_bond`] are keyless

I really can't think of a nice way to check/enforce this but maybe something to think about for FRAME. I'd say au that not having a way to ensure/check that the origin is a keyless account may become problematic.

Reply:
> I don't know of a way to check if an account is keyless

Was thinking of the same. The only thing is that if we create a one-off sub-account with a "secure random derivation path" from the origin? 

in any case this is a problem of the future.